storage/innobase/buf/buf0buf.cc

```diff
@@ -1878,6 +2095,96 @@ inline void buf_pool_t::resize()
 	return;
 }
 
+static std::mutex garbage_collect_mutex{};
+
+inline void buf_pool_t::garbage_collect()
+{
+  std::unique_lock<std::mutex> lk(garbage_collect_mutex);
```

Comment:
I don’t think we need an extra mutex here. If the adaptive hash index is not being used on the pages that are being freed here, the entire critical section will be covered by `buf_pool.mutex`. Else, the `buf_pool.LRU` traversal is protected by the combination of `buf_pool.mutex` and `buf_pool.lru_hp`.

Reply:
its covering the RQG tests where multiple triggers of garbage_collect can occur the same time. I wasn't sure if this was sufficiently protected. 

Could be under a debug build only if needed.

Reply:
I do not think that multiple concurrent calls to `buf_pool.garbage_collect()` should be an issue. Each of them will attempt a scan of the buffer pool. If we have `block->index` for any of the evicted pages (the AHI is disabled by default), then in the worst case, each iteration will `goto rescan` several times. Progress should be guaranteed even in that case.

In the normal case, one of the threads at a time will acquire and hold `buf_pool.mutex` for the entire duration of the iteration. Basically, the `garbage_collect_mutex` is duplicating the role of `buf_pool.mutex`.